signature.1
.\" signature.1 - the dynamic signature program
.\"
.\" $Id: signature.1,v 1.1 2009/10/24 17:48:09 reynolds Exp $
.\"
.\" Copyright (C) Ian Macdonald <ian@caliban.org>
.\"
.\" Process this file with
.\" groff -man -Tascii signature.1
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

.TH SIGNATURE 1 "2nd July 2000" "Ian Macdonald"

.ds sg \fBsignature\fP

.SH NAME
signature \- the dynamic signature program

.SH SYNOPSIS
.B signature
[
.B -d
] [
.B -f
.I FIFO
] [
.B -n
] [
.B -p
.I program
|
.B -q
.I quote file
] [
.B -r
] [
.B -t
.I template
]
.br
.B signature -h
|
.B -v

.SH DESCRIPTION
\*(sg is a program for producing dynamic signatures for e-mail
and news postings.

.SH OPTIONS
.B
.IP -d
print debugging messages on stderr

.B
.IP "-f \fIFIFO\fR"
Use this file as the named pipe. This defaults to
.IR ~/.signature ,
which is usually what you want.

.B
.IP -h
Display this help message and exit.

.B
.IP -n
Don't fork into background as daemon.

.B
.IP "-p \fIprogram\fR"
Use this program instead of
.BR fortune (6) .
Your $PATH will be searched for the program, which can can be supplied with
arguments by quoting the entire string supplied to this option.

.B
.IP "-q \fIquote file\fR"
Use this quote file instead of calling
.BR fortune (6) .

.B
.IP -r
Force removal of an existing signature FIFO.
By default, \*(sg will fail with an error message if it detects
that the file it wishes to use as a FIFO already exists.

.B
.IP "-t \fItemplate\fR"
Use this signature template instead of default.

.B
.IP -v
Display version information and exit.

.SH FILES
.I ~/.signature
.RS
The default FIFO used for passing signatures.

.SH ENVIRONMENT
HOME
.RS
Used for ascertaining the user's home directory, if the default
FIFO is used.
.RE
PATH
.RS
Used for searching for external programs supplied to the
.B -p
option.
.RE

.SH NOTES
.SS SIGNATURE TEMPLATES
The default signature template looks something like this:

\fC
.nf
Linux, the choice          | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
of a GNU generation   -o)  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Kernel %-9.9s          /\\  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
on a %-6.6s           _\\_v | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
                           | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
.fi

.PP
If you want to make your own template, you should create a plain
ASCII file containing the desired fixed text and placeholders.
The area with < characters denotes where the dynamic signature
text should be placed. Currently, there should be no more than
one of these areas per line.

.PP
If printf style format specifiers are encountered, the first of
these will be replaced with the current kernel version. The second
will be replaced with the current architecture.

.SS QUOTE FILES
The format of a quote file is something like this:

\fC
.nf
There is no sincerer love than the love of food.
                -- George Bernard Shaw
~
There is no such thing as a problem without a gift for you in its hands.
~
Do infants have as much fun in infancy as adults do in adultery?
~
"I wanna tell you a tale,
 of how I loved and
 how I failed"
 -- The Verve
~
.fi

.PP
In other words, a quote file should consist of one or more quotes,
separated by a tilde on a line by itself. The last quote should also be
terminated by a tilde.

.PP
As an alternative to the tilde, the percent sign (%) may also be used.

.SH BUGS
.PP
It is not possible to preserve the white space formatting of a quote
file quote or fortune program cookie.

.PP
Strange things may happen if the external program used to generate a
signature produces % signs in its output. This may be interpreted as a
printf(3) format specifier.

.PP
Pine fails to play properly with signature, because an fstat(2) on the
signature FIFO prior to opening it returns 0 in stat.st_size, causing
Pine to open and close it without actually reading anything. This
causes a SIGPIPE, which signature calmly ignores.

.PP
E-mail bug reports to the author. Better yet, send a diff.

.SH AUTHOR
Ian Macdonald <ian@caliban.org>

.SH SEE ALSO
.BR printf (3),
.BR fortune (6)


NEWS
Current Version: 0.14
See the file ChangeLog for the details of all changes.

User-visible changes from 0.13 to 0.14

* fixed bug that caused '<' placeholder to remain behind when there were
  backspaces or underscores in the text

User-visible changes from 0.12 to 0.13

* white space at the ends of lines is now trimmed
* ^H (backspace) and _ characters are now ignored in fortune text
* added a couple of debugging messages (we now report PID and FIFO name)
* fixed bug that caused FIFO and quote file not to be found when using -f and -q

User-visible changes from 0.11 to 0.12

* signature no longer exits when used with Pine. It still doesn't work,
  but it no longer exits either.

User-visible changes from 0.10 to 0.11

* external programs can now be called with arguments (shell quoting required)
* quote files can now use either ~ or % as a separator

User-visible changes from 0.04 to 0.10

* no-daemon mode is now selected with -n
* -d now provides debugging output on stderr
* support for quote files with -q
* bug fix for some fatal errors not clearing up temporary FIFO
* man page updated for quote files

User-visible changes from 0.03 to 0.04

* configure now checks for availability of offensive fortunes
  (Red Hat doesn't use them by default)
* man page updated to reflect the above

User-visible changes from 0.02 to 0.03

* when supplying an alternative program to fortune with the -p switch,
  the full path no longer has to be provided
* man page updated to reflect the above change
* usage message omitted -f in option list
* signature now forks into background on start-up. The -d switch has
  been added to prevent this for debugging purposes.

User-visible changes from 0.01 to 0.02

* default sig format string changed for kernel and architecture strings
  of different lengths
* zombie process build-up bug solved
* usage message tidied up


signature.c
#define _XOPEN_SOURCE

#include <config.h>

#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "signature.h"
#include "quote.h"
#include "format.h"

/* name of default signature file */

#define DEFAULT_FIFO ".signature"

/* *INDENT-OFF* */

/* text of default signature */
#define SIGNATURE							      \
"Linux, the choice          | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n"  \
"of a GNU generation   -o)  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n"  \
"Kernel %-9.9s       /\\  | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n"    \
"on a %-6.6s           _\\_v | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n" \
"                           | <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<\n"

/* *INDENT-ON* */

static char		fifo[ PATH_MAX + 1];
static char		quote_file[BUF_SIZE];
unsigned		debug = 0;

/* display version information */

void
version(
	void
)
{
    printf( "%s %s (%s %s)\n", PACKAGE, VERSION, __DATE__, __TIME__ );
    exit(EXIT_SUCCESS);
}

/* display help message */

void
help(
	int		status
)
{
	static char const * const	msgs[] =	{
		"usage: signature [-d] [-f FIFO] [-h] [-n] [-p program |-q qfile] [-t template] [-v]",
		"-d              print debugging messages on stderr",
		"-f <FIFO>       FIFO instead of ~/.signature",
		"-h              display this help message",
		"-n              don't fork into background as daemon",
		"-p <program>    use this program instead of fortune(6)",
		"-q <qfile>      quote from this file instead of fortune(6)",
		"-t <template>   use signature template instead of default",
		"-v              display version information",
	};
	static size_t		Nmsgs = sizeof( msgs ) / sizeof( msgs[0] );
	char const * const * const lmp = msgs + Nmsgs;
	char const * const *	mp;

	for( mp = msgs; mp < lmp; ++mp )	{
		printf( "%s\n", *mp );
	}
	exit(status);
}

/* remove the FIFO */

void
unlink_fifo(
	void
)
{
	if( debug )	{
		fprintf( stderr, "Unlinking FIFO...\n" );
	}
	unlink( fifo );
}

/* when we catch a signal */

void
caught_signal(
	int		sig
)
{
	if( debug )	{
		fprintf( stderr, "Caught signal %d. Clearing up...\n", sig );
	}
	unlink_fifo();
	exit( EXIT_SUCCESS );
}

/* ignore SIGPIPE */

void
ignore_sigpipe(
	int		sig
)
{
	if (debug) {
		fprintf(stderr, "Caught a SIGPIPE. Using Pine?\n");
	}
}

/* set up FIFO */

void
init_fifo(
	void
)
{
	unlink( fifo );
	if (mkfifo(fifo, 0644) == -1) {
		fprintf(stderr, "Couldn't create %s as FIFO\n", fifo);
		exit(EXIT_FAILURE);
	}
}

/* install signal handler */

void
install_sig_handler(
	void
)
{
	struct sigaction	act;

	/* Install signal handler					 */
	act.sa_handler = caught_signal;
	sigemptyset( &act.sa_mask );
	act.sa_flags = 0;
	sigaction( SIGINT, &act, 0 );
	sigaction( SIGHUP, &act, 0 );
	sigaction( SIGTERM, &act, 0 );
	act.sa_handler = ignore_sigpipe;
	sigemptyset( &act.sa_mask );
	sigaction( SIGPIPE, &act, 0 );
}

int
main(
	int		argc,
	char *		argv[]
)
{
	char		signature[BUF_SIZE+1];
	char		sig_buf[BUF_SIZE+1];
	char		fortune[BUF_SIZE+1];
	char *		producer = NULL;
	char *		prodarg = NULL;
	char *		prodargs[MAX_PROG_ARGS+1];
	char *		slash_ptr;
	char *		cwd_ptr;
	int		prodargc = 0;
	unsigned short int no_fork = 0;
	int		exit_status = 0;
	int		sig_fd;
	int		sig_buf_len = 0;
	int		opt;
	int		pid;
	int		file_pipes[2];
	int		fifo_fd;
	int		num_read;
	int		total = 0;
	FILE *		quote_stream;
	struct sig_node * first_node;
	struct sig_node * * first_node_ptr = &first_node;

	memset( &quote_file, 0, BUF_SIZE );
	/* Get option switches						 */
	while( ( opt = getopt( argc, argv, "df:hnp:q:t:v" )) != -1 )	{
		switch( opt )	{
		case 'd':
			/* Set debugging flag				 */
			debug = 1;
			break;
		case 'f':
			/* Get location of FIFO				 */
			slash_ptr = strrchr( optarg, '/' );
			if( slash_ptr == NULL || *optarg == '.' )	{
				/* Make path absolute			 */
				cwd_ptr = getcwd( NULL, BUF_SIZE );
				sprintf( fifo, "%s/%s", cwd_ptr, optarg );
				free( cwd_ptr );
			} else	{
				strcpy( fifo, optarg );
			}
			if( debug )	{
				fprintf( stderr, "FIFO is %s\n", fifo );
			}
			break;
		case 'h':
			help( EXIT_SUCCESS );
			break;
		case 'n':
			no_fork = 1;
			break;
		case 'p':
			if( quote_file[0] != '\0' )	{
				help( EXIT_FAILURE );
			}
			producer = optarg;
			prodargs[prodargc++] = strtok( producer, " " );
			while( ( prodarg = strtok( NULL, " " )) != NULL )	{
				prodargs[prodargc++] = prodarg;
				if( prodargc > MAX_PROG_ARGS )	{
					fprintf( stderr, "More than %d arguments to %s. "
					"Excess arguments ignored...\n",
					MAX_PROG_ARGS, producer );
					break;
				}
			}
			prodargs[prodargc] = NULL;
			/*
			 * check for existence of program by forking and
			 * then trying to exec() it in the child
			 */
			pid = fork();
			switch( pid )	{
			default:
				/* We are the parent			 */
				waitpid( pid, &exit_status, 0 );
				if( exit_status != EXIT_SUCCESS )	{
					fprintf( stderr, "Program %s doesn't seem to exist\n",
					producer );
					exit( EXIT_FAILURE );
				}
			case -1:	/* Oh well			 */
				perror("Couldn't fork() a child process" );
				exit( EXIT_FAILURE );
			case 0:		/* We are the children		 */
				/* Close stdout				 */
				close( 1 );
				execlp(producer, producer, (char *) 0 );
				exit( EXIT_FAILURE );
			}
			break;
		case 'q':
			if( producer )	{
				help( EXIT_FAILURE );
			}
			slash_ptr = strrchr( optarg, '/' );
			if( slash_ptr == NULL || *optarg == '.' )	{
				/* Make path absolute			 */
				/* Get the current directory		 */
				cwd_ptr = getcwd( NULL, BUF_SIZE );
				/* Get absolute path of quote_file	 */
				sprintf( quote_file, "%s/%s", cwd_ptr, optarg );
				free( cwd_ptr );
			} else	{
				/* Get a quote from an external file instead of fortune */
				strcpy( quote_file, optarg );
			}
			break;
		case 't':
			if( debug )	{
				fprintf( stderr, "Opening template file %s...\n", optarg );
			}
			sig_fd = open( optarg, O_RDONLY );
			if( sig_fd == -1 )	{
				fprintf( stderr, "Couldn't open template %s for reading\n",
				optarg );
				exit( EXIT_FAILURE );
			}
			memset( &sig_buf, 0, BUF_SIZE );
			sig_buf_len = read( sig_fd, &sig_buf, BUF_SIZE );
			if( sig_buf_len == -1 ) {
				perror("Failed to read() template file" );
				exit( EXIT_FAILURE );
			}
			close( sig_fd );
			break;
		case 'v':
			version();
			break;
		case ':':
			fprintf( stderr, "Option needs a value\n" );
			break;
		case '?':
			exit( EXIT_FAILURE );
		}
	}
	if( fifo[0] == '\0' )	{
		/* Set up the path to default FIFO			 */
		sprintf( fifo, "%s/%s", getenv( "HOME" ), DEFAULT_FIFO );
	}
	/* Initialise FIFO						 */
	init_fifo();
	/* Install signal handler					 */
	install_sig_handler();

	if( !no_fork )	{
		if( daemon( 0, 0 ) )	{
			perror( "daemon()" );
			unlink_fifo();
			exit( EXIT_FAILURE );
		}
	}

	quote_stream = quote_file[0] ?	index_quote_file(
			quote_file,
			&total,
			first_node_ptr
	) : NULL;

	/* Start main loop						 */
	for( ; ; )	{
		memset( &signature, 0, BUF_SIZE );
		if( sig_buf_len > 0 )	{
			/* Use user-supplied template			 */
			strcpy( signature, sig_buf );
		} else	{
			/* Use stock signature				 */
			strcpy( signature, SIGNATURE );
		}

		if(access( fifo, F_OK ) != 0 )	{
			fprintf( stderr, "%s seems to have disappeared\n", fifo );
			exit( EXIT_FAILURE );
		}

		/* Open FIFO for writing				 */
		fifo_fd = open( fifo, O_WRONLY );
		if( fifo_fd == -1 )	{
			fprintf( stderr, "Couldn't open %s for writing\n", fifo );
			unlink_fifo();
			exit( EXIT_FAILURE );
		}
		/* Create a pipe					 */
		if(pipe( file_pipes ) != 0 )	{
			perror("Couldn't create pipe() for passing signature" );
			unlink_fifo();
			exit( EXIT_FAILURE );
		}
		/* Fork off a child					 */
		pid = fork();
		switch( pid )	{
		case -1:		/* Oh well			 */
			perror("Couldn't fork() a child process" );
			unlink_fifo();
			exit( EXIT_FAILURE );
		case 0:			/* In child			 */
			/* Close stdout					 */
			close( 1 );
			/* Duplicate write fd as stdout			 */
			dup( file_pipes[1] );
			/* Close write fd				 */
			close( file_pipes[1] );
			/* Close read fd				 */
			close( file_pipes[0] );

			/* Check if we need to print from a quote file	 */
			if( quote_file[0] != '\0' )	{
				exit(print_quote( quote_stream, total, first_node ));
			} else if( !producer )	{/* Call fortune	 */
				if( debug )	{
					fprintf( stderr, "Calling %s...\n", FORTUNE );
				}
				/* Send fortune cookie to write fh	 */
				execl( FORTUNE, "fortune", "-s"
# ifdef OFFENSIVE_FORTUNES
					"a"
# endif /* OFFENSIVE_FORTUNES */
				, (char *) 0 );
				/* We only reach this point if exec() didn't work */
				fprintf( stderr, "Couldn't exec %s\n", FORTUNE );
				exit( EXIT_FAILURE );
			} else	{	/* External program		 */
				if( debug )	{
					fprintf( stderr, "Calling %s...\n", producer );
				}
				/* Send producer's output to write fh	 */
				execvp( producer, prodargs );
				/* We only reach this point if exec() didn't work */
				fprintf( stderr, "Couldn't exec %s\n", producer );
				exit( EXIT_FAILURE );
			}
			exit( EXIT_FAILURE );
			/* NOTREACHED					 */
		default:		/* In parent			 */
			if( debug )	{
				fprintf( stderr, "Child process PID %d...\n", pid );
			}
			/* Clean up after child				 */
			waitpid(pid, (int *) 0, 0 );
			/* Close write fd				 */
			close( file_pipes[1] );
			/* Read fortune cookie through pipe		 */
			num_read = read( file_pipes[0], fortune, PIPE_BUF );
			if( !num_read ) {
				/* Exec() didn't work: quit parent	 */
				perror("Error in read() of FIFO" );
				unlink_fifo();
				exit( EXIT_FAILURE );
			}
			/* Close read fd				 */
			close( file_pipes[0] );
			/* Do all the hard stuff			 */
			format_sig( signature, fortune, num_read );
			/* Send it to the FIFO				 */
			write(fifo_fd, signature, strlen( signature ));
			/* Close FIFO					 */
			close( fifo_fd );
		}
		/* Loop for next signature				 */
		sleep( 1 );
	}
	exit( 0 );
}
